docs/lang/articles/kernels/kernel_function.md

```diff
@@ -95,26 +104,98 @@ You call a kernel either directly or from inside a native Python function. You m
 
-- Scalars and `ti.Matrix` are *passed by value*.
-- `ti.types.ndarray()` and `ti.template()` are passed by reference. This means that any modification to the arguments in the kernel being called also affects the original values.
+- The supported types are defined in the `ti.types` module (see the [Type System](../type_system/type.md) for more information).
+
+:::
 
-> We skip `ti.template()` here and leave it to a more advanced topic: [Metaprogramming](../advanced/meta.md#template-metaprogramming).
+#### Primative Types
+Primative Types are predefined by the python language. This includes types such as Integers, Floats, and Booleans.
 
-In the following example, the arguments `x` and `y` are passed to `my_kernel()` *by value*:
+In the following example, the arguments `x` and `y` are passed to `myKernel()` *by value*:
 
 ```python {1}
 @ti.kernel
-def my_kernel(x: int, y: float):
+def myKernel(x: int, y: float):
     print(x + y)
 
-my_kernel(1, 1.0)  # Prints 2.0
+myKernel(1, 1.0)  # Prints 2.0
+```
+
+#### Vectors/Matrices
+In addition to primative types, we can also define predefined arrays and pass them into the kernel function. Vectors are represented as arrays and Matrices are represented as a group of arrays.
+
+In the following example, we pass a 2x3 matrix `arr` into `myKernel()` *by reference*:
+```python {13,17}
+import taichi as ti
+ti.init()
+
+
+@ti.kernel
+def myKernel(arr: array):
+    #get the number at position arr[1][1]
+    for i in range(len(arr)):
+        for j in range(len(arr[i])):
+            if i == 1 and j == 1:
+                return arr[i][j]
+
+arr = [[2,4,6], [8,10,12]]
+print(myKernel(arr))  #will return 10
+
+```
+#### Struct/Dataclass Types
+We can take advantage of Taichi's type system and create our own dataclass to pass into the kernel function.
+
+In the following example, we created a Rectangle class `Rectangle` and passed an instance of the class `a` into the kernel function to calculate its area.
+
+```python
+import taichi as ti
+@ti.kernel
+def myKernel(a: Rectangle):
+    return a.length * a.height
+
+@ti.dataclass
+class Rectangle:
+    length: float
+    width: float
+
+rectangle = Rectangle(2,2)
+print(myKernel(rectangle))  #will print 4
+
+```
+
+#### Matrix Fields
+We can use the ti.Matrix() function to declare our own [Matrix Field](https://docs.taichi-lang.org/docs/master/field#matrix-fields) and pass it into the kernel. This also works for [Vector Fields](https://docs.taichi-lang.org/docs/master/field#vector-fields)
+
+In the following example, we declare a Matrix Field, generate its values, and pass it into `myKernel()`:
+
+```python
+import taichi as ti
+ti.init()
+
+
+@ti.kernel
+def myKernel(a: ti.Matrix.field):
+    for i in ti.grouped(a):
+        a[i] = [[1,1,1], [1,1,1]]
+
+a = ti.Matrix.field(n=2, m=3, dtype=ti.f32, shape=(2, 2))  #Declares a 2x2 matrix field, with each of its elements being a 3x2 matrix
+myKernel(a)
+print(a[0][0,0])  #prints 1
+
 ```
```

Comment:
```suggestion
```
I'm sorry, `Matrix.Field` is a function call, not a valid type hint. We should better not mention it here.